### DISCUSSION TITLE
Discussion: When to initialize Proposals

### DISCUSSION THREAD

**Discussion by jackhack00** [jackhack00/aztec-gddt#81] [General]
In `p_select_proposal` we pull proposals for the current process by `proposals = state['proposals'].get(process.uuid, [])`.
For that, we need to first instantiate proposals for each process, or instantiate proposals with another policy function in same process. 
Since we need the active sequencers, which changes after each process, I think we need:

`p_init_proposals`
Needed:  
* add `score` to sequencer class
* sort sequencers in interacting users (meaning sort every sequencer with stake_amount over min_stake) and init a proposal for the x highest scores in there, using their score as the proposal score. 

When/Where:
* In Pre-Phase, after updating interacting_users

What this helps with:
* Separate init proposals from selecting proposals
* Introduce 'score' to Sequencer, allowing us to (in v2) separate the concept of a sequencer score (which is essentially their pubkey used for vrf) and proposal score (which might later take other variables into account)
* Lets us cleanly select x of y proposals, without initializing proposals for all interacting users. 

Alternatively:
* Pack init logic into p_select_proposal

**Comment by danlessa** [comment]
Note: When/Where should be during `SelectionPhase.pending_proposals`

**Comment by andrew-octopus** [comment]
I have put a candidate solution in https://github.com/BlockScience/aztec-gddt/pull/25. Currently there is no check to see if the `pending_proposals` criterion is met, but this should be an easy assert statement. If the Pull Request is approved, please close this Issue.
